### ISSUE TITLE
Make RegexParseException public

### ISSUE THREAD

**Issue by danmosemsft** [dotnet/runtime#1080]
Right now when regexes fail to parse, we throw an ArgumentException with a string. However internally we have more information -- the offset into the pattern and in many cases the type of failure eg `RegexParseError.ReversedCharRange`. These are included in most or all cases in the exception, but only when cast to RegexParseException which is an internal type. 

As noted here https://github.com/dotnet/corefx/pull/29178#discussion_r182842595 and https://github.com/dotnet/runtime/issues/372#issuecomment-559245612 tooling could potentially use this data eg to squiggle the location in a regex. 

(Not apparently @CyrusNajmabadi's Roslyn parser by that's used by VS -- although it obtains the offset by essentially reimplementing the parser, it apparently has other reasons it had to do this.)

If we did this we would have to be comfortable with sometimes changing the offset and/or error for particular invalid patterns, since this may be inevitable when changing and refactoring hte parser. We would not want to consider that a breaking change.

I have no opinion on whether this is worthwhile, just logging to record it.

**Comment by stephentoub** [comment]
@danmosemsft, unless there are known scenarios demanding this, I suggest it be closed. It's otherwise just going to sit open for a long time.

**Comment by danmosemsft** [comment]
Anyone who thumbed this up want to make a case for having this? If not we will close it.

**Comment by abelbraaksma** [comment]
@danmosemsft, @stephentoub, actually, yes, I do like to make a strong case for this. I already use the internal error by hacking into it with reflection (it's been a while ago, I'd have to check what I did exactly).

My use case is a compiler and I need to give detailed, and sometimes altered information to the user. That's because the regex he gives is in a slightly different syntax (W3C syntax for XPath 3.1), and it must support surrogate pairs (which are totally unsupported in .NET regexes: try any class, or negated class, it will not properly match surrogate pairs as the single Unicode codepoint it represents). I modify the regex to properly support this (i.e., `.{3}` gets rewritten such that a surrogate pair is a single char, and much more). In case of errors, I need to translate the position etc back to the original string. 

Also, it's just easier to read the detailed exception data and create a customized and localized message, than it is by parsing the returned exception message, which can be disastrous in different locales).

Another use case may be simply that source regexes are given as part of a script (in my case, they're inside XSLT 3.0 stylesheets, for which I've written a compiler), and that you need to calculate the proper offset of the error in the string.

Yet another situation is where a regex appears from a configuration file in, say, XML in an attribute. The user may have put newlines in the regex for readability, but the XML parser will collapse all whitespace for attributes. As with previous examples, in such cases the offset of the error needs to be adjusted to take this into account to give the user proper info.

**Comment by abelbraaksma** [comment]
> If we did this we would have to be comfortable with sometimes changing the offset and/or error for particular invalid patterns, since this may be inevitable when changing and refactoring hte parser. We would not want to consider that a breaking change.

Anyone currently relying on this (like me) already expects that these internals can change. I don't believe, btw, that there is a MS policy that exceptions and there messages or data are required to remain consistent between releases. In the F# team I've seen many changes to existing errors age texts without that being considered breaking. 

Of course, the type itself should remain binary compatible, but that shouldn't be too hard.

**Comment by danmosemsft** [comment]
Either of you want to propose the public api? Would it include the enum?

**Comment by abelbraaksma** [comment]
I can certainly give it a go. At first glance, I think exposing the enum would be fine. We shouldn't be too afraid to add cases to it if future changes warrants that (I think that's within the range of compat guidelines).

**Comment by danmosemsft** [comment]
There is some info about API proposals here https://github.com/dotnet/runtime/blob/master/docs/project/api-review-process.md

**Comment by danmosemsft** [comment]
Before making it public probably most of the work would be to ensure that RegexParseError was always populated. If you look at RegexParserTests.cs, you will see many cases where they verify null,  because there was no concerted effort to consistently set a parse error. In fact, this work should probably happen before making the API proposal, as it will help validate that the RegexParseError enum has the right set of values.

**Comment by abelbraaksma** [comment]
I had a quick look at the code, but I couldn't immediately verify your assessment, so I'll set up a test environment in VS and open the regex solution.

**Comment by danmosemsft** [comment]
@abelbraaksma my bad: all the error cases seem to set a RegexParseError - in fact, the RegexParseException constructor requires it. I guess you'd still want to satisfy myself that RegexParseError has the best set of values.

**Comment by abelbraaksma** [comment]
> the RegexParseException constructor requires it

I noticed that, so I assumed you meant something like places where an `ArgumentException` was raised instead, which wouldn't have the internal field set that's used in testing. 

The set of values appears pretty exhaustive, but I wouldn't be surprised if, while work is still ongoing on the performance, that new cases emerge for further fine tuning.

**Comment by danmosemsft** [comment]
@abelbraaksma just to be sure, are you waiting on me, or are you going to write up a brief proposal?  no obligation.

**Comment by abelbraaksma** [comment]
@danmosemsft, sorry for the delay, I got distracted and kinda forgot, thanks for the ping. I'll come up with a proposal soon :)

(doesn't github have a way of assigning tasks or something, btw, such that I could have a TODO list in my profile?)

**Comment by danmosemsft** [comment]
Assigned! You're on your own schedule

**Comment by danmosemsft** [comment]
@abelbraaksma any update?

**Comment by abelbraaksma** [comment]
@danmosemsft, oops, you're right! I'll set myself to work on this, got a little too involved with F# repo recently. You can expect a PR in a few hours :)

**Comment by danmosemsft** [comment]
@abelbraaksma to be clear, we're not ready for a PR yet. The next step is to create an API proposal issue, per guidance here:  https://github.com/dotnet/runtime/blob/master/docs/project/api-review-process.md
There's actually a template for you: https://github.com/dotnet/runtime/issues/new?assignees=&labels=api-suggestion&template=02_api_proposal.md&title=

You can either stick the text here and I'll update the top post of this issue, or create a new issue and I"ll close this one in favor of yours.

Then, as noted in the doc, it can take a while before it gets to API review as there's a lot waiting right now (https://github.com/dotnet/runtime/issues?q=is%3Aopen+is%3Aissue+label%3Aapi-ready-for-review). After it gets api-approved (assuming it's not rejected) we can take a PR either from you or someone else.

I realize that's not as fun as getting the code in right away 😄

**Comment by abelbraaksma** [comment]
Thanks for the extra guidance. I meant with "PR" submitting an RFC (not yet the code change), which is the common way in the F# repo. Here it is done through a tagged issue, indeed I wasn't aware of that.

The proposal is there, thanks for your patience: https://github.com/dotnet/runtime/issues/38872

**Comment by eerhardt** [comment]
Do we need both #38872 and this issue? Or can this one be closed in favor of #38872?

**Comment by abelbraaksma** [comment]
@eerhardt, I'd suggest to close this in favor of the new one.
